src/solvers/td3.ml

```diff
@@ -154,6 +156,9 @@ module WP =
       let abort_verify = GobConfig.get_bool "solvers.td3.abort-verify" in
       let prev_dep_vals = HM.create 10 in
 
+      (* Tracks dependencies between an unknown and the things it depends on AND has side-effects to *)
+      let dep = data.dep in
```

Comment:
It seems like side-effects are not always recorded into `side_infl`? Is there some specific reason for that?

Reply:
Well, before this `side_infl` was only populated during postsolving for marshaling, but wasn't needed to collect during the solving itself.
Of course that can be changed, but I'm wondering if that's even necessary since leaves (side effect targets with no right-hand sides) themselves don't have warnings, which are emitted during their rhs evaluation. So there shouldn't even be anything to reuse from them in `var_messages` and `rho_write`.

Reply:
None of the tests seem to fail but I am still a bit skeptical about using `side_infl`: it is explicitly cleared before we start the solver, which strikes me as very odd.

It seems that we would run the risk of removing globals that are only side-effected to from superstable things (but remain reachable) if we remove the things that side-effects are done to from `dep` (though I could not construct an example thus far).

~~~OCaml
        (* reachability will populate these tables for incremental global restarting *)
        HM.clear side_dep;
        HM.clear side_infl;
~~~

Reply:
Maybe the better solution is to rename `dep` to `uses` or something that indicates that these are the unknowns that `x` reads or has side-effects to?

Reply:
> None of the tests seem to fail but I am still a bit skeptical about using `side_infl`: it is explicitly cleared before we start the solver, which strikes me as very odd.

Right, the previous use of `side_infl` was just for restarting, which is all done before starting the solving itself. So to minimize memory usage and not keep around data during the solving, which will be recollected during postsolving anyway.

> It seems that we would run the risk of removing globals that are only side-effected to from superstable things (but remain reachable) if we remove the things that side-effects are done to from `dep` (though I could not construct an example thus far).

Hmm, you may have some point here, but if that's indeed a problem, then it's broader than just this PR. Maybe it becomes a problem if at least two incremental changes are made (so there's an intermediate run, which both loads and saves). Then I suppose that might lose the original data and not be able to perform restarting on the final execution correctly?

Since all the testing of incrementality has always been just one change, such an issue has never had a chance to reveal itself. I guess @stilscher's benchmarks that do longer chains of incremental changes might also expose these.

Reply:
So maybe this sort of structure is needed indeed and `uses` name with fewer mixup opportunities.

Now I'm wondering if it's even necessary to collect this `uses` during the actual solving or again just during postsolving. The postsolver sees all the changed/new uses and all the `reachable_and_superstable` should also remain.